### ISSUE TITLE
Handling 0-length arrays in JNI Array Critical in GC

### ISSUE THREAD

**Issue by dmitripivkine** [eclipse-openj9/openj9#20074]
The representation of 0-length array `<class><0><0>` is technically Discontiguous. The criteria for Discontiguous "size in elements" field contains 0 and it happen for 0-length array naturally due size 0.

According the spec JNI Array Critical section has to have Contiguous representation. If JVM stores an array in any discontiguous form internally it's data part should be copied to the memory in contiguous form.

Currently GC handles 0-length array naturally as Discontiguous. It means at "Get" side we malloc memory (8 bytes VM header word + 0 bytes reserved for array body, so 8 bytes total) and copy 0 bytes to it. Malloc'ed memory is wrapped with Memory Tags for header and footer of course. At "Release" size we copy 0 bytes from malloc'ed memory back to the heap and free malloc'ed memory.
These actions are not necessary for 0-length arrays and can be avoided. We can handle 0-length arrays as Contiguous.

The motivation for this change is not only code simplification but also preparing for coming off-heap changes.

The length of data part for 0-length array is 0 obviously. Despite API returns starting address for Critical Section (points to malloc'ed memory after VM header word) nothing should be written to it. So, currently by returning an address in malloc'ed memory we have control that nothing is written to the memory - footer memory tag should not to be corrupted. By switching to contiguous path the returned address is going to be from object heap range. We can get less noticeable heap corruption potentially. I think this is only a factor we can take to consideration to not switch to contiguous array handling path.

**Comment by dmitripivkine** [comment]
@amicic @gacholio @tajila Did I miss something?

**Comment by gacholio** [comment]
Did I misunderstand this? Are you simply proposing a change to how critical works (don't malloc 0-length arrays) rather than changing the heap shape of 0-length arrays?

**Comment by dmitripivkine** [comment]
Yes, just return heap address instead of malloc'ed. I know, it is trivial probably. Just to be sure I have not missed something.
In general we like an idea to change 0-length arrays format and treat them as contiguous but there is significant JIT work required, so looks like we do not have resources to do this now.

**Comment by gacholio** [comment]
A 0-length contiguous array would be indistinguishable from a discontiguous array (of any size). Ideally, we should hand out a pointer which will crash if written to, but we would need to do something like reserving a page and making it no access. This could be done in the future if desired.

**Comment by dmitripivkine** [comment]
> A 0-length contiguous array would be indistinguishable from a discontiguous array (of any size). Ideally, we should hand out a pointer which will crash if written to, but we would need to do something like reserving a page and making it no access. This could be done in the future if desired.

Currently we do have run time selection of two different Indexable object headers:
- For Standard Collectors Contiguous object header format is `<class><size>` and there is no Discontiguous to be used. So, 0-length array can look like `<class><0>`, there is no need to distinguish it from Discontiguous.
- For Balanced Collector Contiguous object header format is `<class><size><dataAddr>`. If Off-heap is enabled, there is no Discontiguous also. If Off-heap is disabled we can use `dataAddr` to distinguish from Discontiguous.

**Comment by dmitripivkine** [comment]
I noticed that in off-heap prototype we are handling 0-length arrays in heap as contiguous. This change of behaviour can be done separately to reduce risk during switching to off-heap code. Also in general looks like unnecessary to malloc/free memory for nothing mostly.
